src/connectedk8s/azext_connectedk8s/custom.py

```diff
@@ -1015,3 +1065,923 @@ def upgrade_agents(cmd, client, resource_group_name, cluster_name, kube_config=N
+
+    if enable_azure_rbac:
+        if (aad_client_id is None) or (aad_client_secret is None):
+            telemetry.set_user_fault()
+            telemetry.set_exception(exception='Client ID or secret not provided for AAD RBAC', fault_type=consts.Client_Details_Not_Provided_For_AAD_RBAC_Fault,
+                                    summary='Both aad authorization client id and client secret is required to enable AAD RBAC feature')
+            raise CLIError("Please provide both aad authorization client id and client secret to enable AAD RBAC feature")
+
+    if enable_cl:
+        enable_cl, custom_locations_oid = check_cl_registration_and_get_oid(cmd)
+        if not enable_cluster_connect and enable_cl:
+            enable_cluster_connect = True
+            features.append("cluster-connect")
+            logger.warning("Enabling 'custom-locations' feature will enable 'cluster-connect' feature too.")
+        if not enable_cl:
+            features.remove("custom-locations")
+            if len(features) == 0:
+                raise CLIError("Based on the inputs given and the warnings displayed above, there are no features to be enabled.")
+
+    # Send cloud information to telemetry
+    send_cloud_telemetry(cmd)
+
+    # Setting kubeconfig
+    kube_config = set_kube_config(kube_config)
+
+    # Checking whether optional extra values file has been provided.
+    values_file_provided, values_file = utils.get_values_file()
+
+    # Validate the helm environment file for Dogfood.
+    dp_endpoint_dogfood = None
+    release_train_dogfood = None
+    if cmd.cli_ctx.cloud.endpoints.resource_manager == consts.Dogfood_RMEndpoint:
+        dp_endpoint_dogfood, release_train_dogfood = validate_env_file_dogfood(values_file, values_file_provided)
+
+    # Loading the kubeconfig file in kubernetes client configuration
+    load_kube_config(kube_config, kube_context)
+    configuration = kube_client.Configuration()
+
+    # Checking the connection to kubernetes cluster.
+    # This check was added to avoid large timeouts when connecting to AAD Enabled AKS clusters
+    # if the user had not logged in.
+    check_kube_connection(configuration)
+
+    # Get kubernetes cluster info for telemetry
+    kubernetes_version = get_server_version(configuration)
+
+    # Checking helm installation
+    check_helm_install(kube_config, kube_context)
+
+    # Check helm version
+    helm_version = check_helm_version(kube_config, kube_context)
+    telemetry.add_extension_event('connectedk8s', {'Context.Default.AzureCLI.HelmVersion': helm_version})
+
+    # Check for faulty pre-release helm versions
+    if "3.3.0-rc" in helm_version:
+        telemetry.set_user_fault()
+        raise CLIError("The current helm version is not supported for azure-arc onboarding. Please upgrade helm to a stable version and try again.")
+
+    # Check whether Connected Cluster is present
+    if not connected_cluster_exists(client, resource_group_name, cluster_name):
```

Comment:
We should also check if the cc and rg present in configmap match the parameter values. Otherwise, the customer could have provided a different cc resource & we would still upgrade the cluster which points to a different cc resource

Reply:
enable_features and disable_features are similar to update_agents, I followed the same procedure here too.

Reply:
This check should be present in update_agents as well. We might have missed that earlier. But if we have less time, we can take that as an action item for future since it will not be a frequent scenario

Reply:
Adding that validation in enable and disable_features now.

Reply:
Added in enable_disable as of now

Reply:
In that case lot of duplicate code. We should optimize.

---
In reply to: [596660175](https://github.com/Azure/azure-cli-extensions/pull/3001#discussion_r596660175) [](ancestors = 596660175)